src/ol/layer/Layer.js

```diff
@@ -381,6 +381,22 @@ class Layer extends BaseLayer {
     this.rendered = false;
   }
 
+  /**
+   * When the renderer follows a layout -> render approach, do the final rendering here.
+   * @param {import('../Map.js').FrameState} frameState Frame state
+   */
+  renderDeferred(frameState) {
```

Comment:
This new method executes `CanvasRenderingContext2D` instructions that were previously recorded on a `ZIndexContext` (see details in the comments there).

When `declutter` is set on a layer, the rendering cycle controlled by the Composite renderer consists of 3 distinct operations:

1. `renderFrame` (Vector) and `postRender` (VectorTile): record `CanvasRenderingContext2D` instructions on a `ZIndexContext` for everything that is not subject to decluttering.
2. `renderDeclutter`: when all layer renderers' `renderFrame` methods were called by the `Composite` renderer, this method will be called for each layer in opposite rendering order, because higher in the layer stack means higher decluttering priority. Like with `renderFrame`,  all instructions that do not collide with previous instructions or are not subject to decluttering will be recorded to a `ZIndexContext`.
3. `renderDeferred`: after the `renderDeclutter` cycle, the recorded `ZIndexContext` instructions will be replayed to the target canvas context. For `Vector` layers, this is done just by calling `renderDeferred` on the executor group.